synapse/handlers/message.py

```diff
@@ -918,6 +918,7 @@ async def create_new_client_event(
         full_state_ids_at_event = None
         if auth_event_ids is not None:
             # If auth events are provided, prev events must be also.
+            # prev_event_ids could be an empty array though.
             assert prev_event_ids is not None
 
             # Copy the full auth state before it stripped down
```

Comment:
> Separately: I'm not sure its correct that the insertion event points to anything else, as we need want federation to ask for the state at the insertion event, rather than any of the state events that it references?
>
> *-- [@erikjohnston](https://github.com/matrix-org/synapse/pull/11243#discussion_r750312188)*

The insertion event is connected to the state chain so the whole historical batch can share the same `state_group`. It also looks nice in the DAG semantically to be able to see the state that authed the batch.

For the federation cases, it seems to work 🤔🤷. `auth_event_ids` probably

Reply:
Hmm, that surprises me a bit. I was very much imagining that the federation code would ask for the state at the insertion event. I guess it probably works because we fetch the state at the start of the chain of state events? And we return the full state, so we then can accept all the state events we add? I don't think that is necessary to make the whole batch have the same state group.

Also, I wonder if that'll make the client UI show the state events when backpaginating, making it look like the state was changed at those points, when actually they were changed much further back?

Reply:
> I don't think that is necessary to make the whole batch have the same state group.

I can play around with it but that seems like something for another PR. It's pretty finicky to get right.


> Also, I wonder if that'll make the client UI show the state events when backpaginating, making it look like the state was changed at those points, when actually they were changed much further back?

Doesn't seem to be a problem for the local origin homeserver with Element probably because they are marked as `outliers`.

I want to test this on a federated homeserver though.

I'm trying to setup a local federated homeserver to see if it's a problem for remote federated servers. Currently working out the hostname/`server_name` problem mapping to a port locally (tips welcome) and basing the config off of the [demo script](https://github.com/matrix-org/synapse/blob/develop/demo/start.sh). I'm attempting to use [`/etc/hosts` and a nginx reverse proxy to `127.0.0.1:8448` (non-tls)](https://www.baeldung.com/linux/mapping-hostnames-ports) . Synapse is configured to be `tls: false` on `8008 and `8448`. This appears to work to get `my.matrix.host` and `other.matrix.host` but doesn't make it happy enough to be able to successfully fetch the room list.

Might just end up creating a Complement test around making sure the state isn't visible between batches ⏩ but would be nice to double-check with real Element about how it looks.

Reply:
> I'm trying to setup a federated homeserver locally to see if it's a problem for remote federated servers.

FYI you can start some preconfigured HSes that federate with each other by running `./demo/start.sh` (and `./demo/stop.sh`)

Reply:
> Also, I wonder if that'll make the client UI show the state events when backpaginating, making it look like the state was changed at those points, when actually they were changed much further back?

> Might just end up creating a Complement test around making sure the state isn't visible between batches ⏩ but would be nice to double-check with real Element about how it looks.

I updated the [Complement test which tests multiple batches over federation to check for historical state](https://github.com/matrix-org/complement/pull/214) and didn't see any problem as well.

Reply:
> I don't think that is necessary to make the whole batch have the same state group.

I created https://github.com/matrix-org/synapse/pull/11487 to add a test around sharing the `state_groups`. Things seem to work just fine if I disconnect the floating state chain from the `insertion` event so we can move forward with that in https://github.com/matrix-org/synapse/pull/11114 after this merges. Both the state chain and the `insertion` event chain can float with no `prev_events`. No issues with `state_groups` or accepting over federation.